common/beerocks/bcl/include/bcl/network/socket_refactorings.h

```diff
@@ -0,0 +1,795 @@
+     *
+     * @return True on success and false otherwise (for example, if it was already closed)
+     */
+    bool close() override
+    {
+        printf("close(%d)\n", m_fd);
+        if (FileDescriptor::invalid_descriptor == m_fd) {
+            return false;
+        }
+
+        int rc = ::close(m_fd);
+        m_fd   = FileDescriptor::invalid_descriptor;
+
+        return (0 == rc);
+    }
+
+private:
+    int m_fd;
+};
+
+/**
+ * Wrapper class around sockadr (structure describing a generic socket address)
+ */
+class SockAddr {
+public:
+    virtual ~SockAddr()                             = default;
+    virtual const struct sockaddr *sockaddr() const = 0;
+    virtual const socklen_t &length() const         = 0;
+
+    struct sockaddr *sockaddr()
+    {
+        return const_cast<struct sockaddr *>(const_cast<const SockAddr *>(this)->sockaddr());
+    }
+    socklen_t &length()
+    {
+        return const_cast<socklen_t &>(const_cast<const SockAddr *>(this)->length());
+    }
+};
+
+/**
+ * Wrapper class around msghdr (structure describing messages sent by `sendmsg` and received by
+ * `recvmsg`)
+ */
+class MsgHdr {
+public:
+    virtual ~MsgHdr()                           = default;
+    virtual const struct msghdr *msghdr() const = 0;
+
+    struct msghdr *msghdr()
+    {
+        return const_cast<struct msghdr *>(const_cast<const MsgHdr *>(this)->msghdr());
+    }
+};
+
+/**
+ * Any object implementing the Socket interface is a FileDescriptor too.
+ * This way we give meaning to the action of "closing the socket".
+ */
+class Socket : public FileDescriptor {
+    virtual int setsockopt(int level, int optname, void *optval, socklen_t optlen) = 0;
+};
+
+/**
+ * Abstract base class for all types of sockets: Raw, UDP, TCP, UDS, ...
+ * This implementation class aggregates a FileDescriptor implementation so it has a file
+ * descriptor. Methods overridden from FileDescriptor interface delegate on the aggregated
+ * implementation.
+ * Derived classes provide the file descriptor obtained with a call to socket(), using different
+ * family, type and protocol parameters.
+ * This class aggregates a FileDescriptor instead of inheriting from one of its implementations to
+ * follow the principle of "Favor Aggregation over Inheritance".
+ * See https://wiki.c2.com/?UseCompositionAndInterfacesWithoutClassInheritance
+ */
+class SocketAbstractImpl : public Socket {
+public:
+    int fd() override { return m_descriptor->fd(); }
+    bool close() override { return m_descriptor->close(); }
+    int setsockopt(int level, int optname, void *optval, socklen_t optlen) override
+    {
+        return ::setsockopt(m_descriptor->fd(), level, optname, optval, optlen);
+    }
+
+protected:
+    /**
+     * Constructor is protected so only derived classes can call it
+     */
+    explicit SocketAbstractImpl(int fd) : m_descriptor(std::make_shared<FileDescriptorImpl>(fd)) {}
+
+private:
+    std::shared_ptr<FileDescriptor> m_descriptor;

```

Comment:
Why is this a shared pointer? There is no way it can ever be used by anything else because it's private and not exported. Also, it's constructed in the constructor where it is a FileDescriptorImpl, so it can't every be replaced by anything else.

So I think we could just make it a real aggregate, i.e.
```suggestion
    FileDescriptorImpl m_descriptor;
```